test/replication-luatest/gh_8433_replication_timeout_test.lua

```diff
@@ -0,0 +1,87 @@
+local luatest = require('luatest')
+local server = require('luatest.server')
+local replica_set = require('luatest.replica_set')
+
+local g = luatest.group('prevote_fail')
+
+local function get_term(server)
+    return server:exec(function()
+        return box.info.election.term
+    end)
+end
+
+g.after_all(function(g)
+    g.replica_set:stop()
+end)
+
+g.before_all(function(g)
+    g.replica_set = replica_set:new({})
+    g.box_cfg = {
+        election_mode = 'candidate',
+        replication = {
+            server.build_listen_uri('server1'),
+            server.build_listen_uri('server2'),
+            server.build_listen_uri('server3'),
+        },
+    }
+
+    g.server1 = g.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = g.box_cfg,
+    }
+
+    g.server2 = g.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = g.box_cfg,
+    }
+
+    g.server3 = g.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = g.box_cfg,
+    }
+
+    g.replica_set:start()
+    g.replica_set:wait_for_fullmesh()
+
+    for _, server in ipairs(g.replica_set.servers) do
+        server:exec(function()
+            box.error.injection.set('ERRINJ_RAFT_BYPASS_CB_ELECTION_START', 1)
+        end)
+    end
+end)
+
+g.test_prevote_fail = function(g)
+    --
+    -- The test is flaky if not launched on debug version. The following
+    -- situation is acceptable and must be bypassed with error injection
+    -- (ERRINJ_RAFT_BYPASS_CB_ELECTION_START - forbids to start election
+    -- in such case):
+    --
+    --   1. The applier to non-leader fails, decreasing the number of
+    --      healthy_count to 2 and cleaning the flag about according replica
+    --      in leader_witness_map;
+    --
+    --   2. raft_sm_election_update_cb is invoked on timer end (the connection
+    --      to a leader is too slow, it didn't managed to send heartbeat in
+    --      replication_timeout), which causes successful start of elections
+    --
+    luatest.tarantool.skip_if_not_debug()
+
+    local leader = g.replica_set:get_leader()
+    leader:exec(function()
+        box.cfg({replication_timeout = 0.1})
+    end)
+
+    local term = 0
+    -- Wait for new leader to be elected
+    luatest.helpers.retrying({}, function()
+        local new_leader = g.replica_set:get_leader()
```

Comment:
I've encountered such failures when running the test many times:
```
[007] not ok 1	prevote_fail.test_prevote_fail
[007] #   ...replication-luatest/gh_8433_replication_timeout_test.lua:81: Actual and expected values are equal: <table: 0x010bedb108> {
[007] #       alias = "server2",
[007] #       args = <table: 0x010bede540> {},
[007] #       box_cfg = <table: 0x010bedb188> {
[007] #           election_mode = "candidate",
[007] #           replication = <table: 0x010bedb208> {
[007] #               "/tmp/t/007_replication-luatest/server1.sock",
[007] #               "/tmp/t/007_replication-luatest/server2.sock",
[007] #               "/tmp/t/007_replication-luatest/server3.sock",
[007] #           },
[007] #       },
[007] #       command = "/Users/s.petrenko/Source/sptnt/test-run/lib/luatest/luatest/server_instance.lua",
[007] #       env = <table: 0x010bf58118> {
[007] #           TARANTOOL_ALIAS = "server2",
[007] #           TARANTOOL_BOX_CFG = "{\"election_mode\":\"candidate\",\"replication\":[\"\\/tmp\\/t\\/007_replication-luatest\\/server1.sock\",\"\\/tmp\\/t\\/007_replication-luatest\\/server2.sock\",\"\\/tmp\\/t\\/007_replication-luatest\\/server3.sock\"]}",
[007] #           TARANTOOL_LISTEN = "/tmp/t/007_replication-luatest/server2.sock",
[007] #           TARANTOOL_WORKDIR = "/tmp/t/007_replication-luatest/server2-FC8cN9d5lvmm",
[007] #       },
[007] #       id = "FC8cN9d5lvmm",
[007] #       instance_id = 1,
[007] #       net_box = <table: 0x010bfa5d60> {
[007] #           _callback = function: 0x010bfa5e10,
[007] #           _fiber = fiber: 143,
< it goes on enumerating all the spaces net.box knows of>
```

Reply:
Indeed, I was able to reproduce such failure on the old test. It's a new one out there, which seems to be stable. However, sometimes test hangs on `before_all` part with never ending `wait_for_fullmesh`. It seems to be related to a bug in proxy. I created ticket for that: https://github.com/tarantool/luatest/issues/298

Reply:
Thanks!